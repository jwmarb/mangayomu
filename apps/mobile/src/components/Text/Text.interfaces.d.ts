import { ButtonColorsTextContrasts } from '@mangayomu/theme';
import { Colors } from '@mangayomu/theme';
import { TextVariants } from '@theme/theme';
import React from 'react';
import { TextProps as NativeTextProps, TextStyle } from 'react-native';

export interface TextProps extends React.PropsWithChildren, NativeTextProps {
  variant?: TextVariants;
  color?: Colors | ButtonColorsTextContrasts;
  align?: TextStyle['textAlign'];
  bold?: boolean;
  italic?: boolean;
  underline?: boolean;
}
