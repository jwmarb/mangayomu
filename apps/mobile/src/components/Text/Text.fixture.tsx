import React from 'react';
import Text, {
  TEXT_ALIGNMENTS,
  TEXT_COLORS,
  TEXT_VARIANTS,
  TextAlignments,
  TextColors,
  TextVariants,
} from './Text';
import { SafeAreaView } from 'react-native-safe-area-context';
import { useSelect, useValue } from 'react-cosmos/client';
import useModeSelect from '../../hooks/useModeSelect';
import { createStyles } from '../../utils/theme';
import useStyles from '../../hooks/useStyles';

const PANGRAM = 'The quick brown fox jumps over the lazy dog';

const styles = createStyles((theme) => ({
  container: {
    flex: 1,
    backgroundColor: theme.palette.background.default,
    justifyContent: 'center',
  },
}));

export default function TextFixture() {
  const [variant] = useSelect('variant', {
    options: TEXT_VARIANTS as unknown as string[],
  });
  const [color] = useSelect('color', {
    options: TEXT_COLORS as unknown as string[],
    defaultValue: 'textPrimary',
  });
  const [text] = useValue('text', {
    defaultValue: PANGRAM,
  });
  const [alignment] = useSelect('alignment', {
    options: TEXT_ALIGNMENTS as unknown as string[],
    defaultValue: 'left',
  });
  const contrast = useModeSelect();
  const style = useStyles(styles, contrast);
  return (
    <SafeAreaView style={style.container}>
      <Text
        color={color as TextColors}
        contrast={contrast}
        variant={variant as TextVariants}
        alignment={alignment as TextAlignments}
      >
        {variant}. {text}
      </Text>
    </SafeAreaView>
  );
}
