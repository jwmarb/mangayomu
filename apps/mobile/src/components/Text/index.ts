export { default } from './Text';
export { AnimatedText } from './Text';
export * from './Text.helpers';
