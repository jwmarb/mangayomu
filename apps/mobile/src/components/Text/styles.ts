import { StyleSheet } from 'react-native';
export const variants = StyleSheet.create({
  h3: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  h2: {
    fontSize: 18,
    fontWeight: 'bold',
  },
  body: {
    fontSize: 14,
    fontWeight: 'normal',
  },
});
