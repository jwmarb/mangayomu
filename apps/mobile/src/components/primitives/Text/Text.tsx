import { TextProps as NativeTextProps, Text as NativeText } from 'react-native';
import React from 'react';
import Animated from 'react-native-reanimated';
import useContrast from '@/hooks/useContrast';
import {
  TextVariants,
  TextAlignments,
  TextColors,
} from '@/components/primitives/types';
import useStyles from '@/hooks/useStyles';
import {
  variants,
  colors,
  alignments,
  decorators,
} from '@/components/primitives/Text/styles';

export type TextProps = NativeTextProps & {
  variant?: TextVariants;
  color?: TextColors;
  alignment?: TextAlignments;
  contrast?: boolean;
  bold?: boolean;
  italic?: boolean;
};

/**
 * Text component to render text.
 */
function Text(props: TextProps, ref: React.ForwardedRef<NativeText>) {
  const {
    variant: variantProp = 'body1',
    color: colorProp = 'textPrimary',
    alignment: alignmentProp = 'left',
    contrast: contrastProp,
    bold,
    italic,
    ...rest
  } = props;
  const contrast = useContrast(contrastProp);
  const color = useStyles(colors, contrast)[colorProp];
  const alignment = alignments[alignmentProp];
  const variant = variants[variantProp];
  const style = [
    variant,
    color,
    alignment,
    bold ? decorators.bold : [],
    italic ? decorators.italic : [],
    rest.style,
  ];
  return <NativeText {...rest} style={style} ref={ref} />;
}

const ForwardedText = React.forwardRef(Text);

export const AnimatedText = Animated.createAnimatedComponent(ForwardedText);

export default ForwardedText;
