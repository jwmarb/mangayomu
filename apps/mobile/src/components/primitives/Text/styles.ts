import { StyleSheet, TextStyle } from 'react-native';
import {
  TextAlignments,
  TextColors,
  TextVariants,
} from '@/components/primitives/types';
import { createStyles } from '@/utils/theme';
export const variants = StyleSheet.create<Record<TextVariants, TextStyle>>({
  h1: {
    fontSize: 48,
    fontWeight: 'bold',
    letterSpacing: 1.68,
  },
  h2: {
    fontSize: 34,
    fontWeight: 'bold',
    letterSpacing: 1.2,
  },
  h3: {
    fontSize: 24,
    fontWeight: '400',
    letterSpacing: -0.16,
  },
  body1: {
    fontSize: 16,
    fontWeight: 'normal',
  },
  body2: {
    fontSize: 14,
    fontWeight: 'normal',
    letterSpacing: 0.44,
  },
  button: {
    fontSize: 16,
    fontWeight: '500',
    letterSpacing: 0.56,
  },
});

export const colors = createStyles<Record<TextColors, TextStyle>>((theme) => ({
  primary: {
    color: theme.palette.primary.main,
  },
  secondary: {
    color: theme.palette.secondary.main,
  },
  textPrimary: {
    color: theme.palette.text.primary,
  },
  textSecondary: {
    color: theme.palette.text.secondary,
  },
  disabled: {
    color: theme.palette.text.disabled,
  },
  error: {
    color: theme.palette.error.main,
  },
}));

export const alignments = StyleSheet.create<Record<TextAlignments, TextStyle>>({
  left: {
    textAlign: 'left',
  },
  center: {
    textAlign: 'center',
  },
  right: {
    textAlign: 'right',
  },
});
