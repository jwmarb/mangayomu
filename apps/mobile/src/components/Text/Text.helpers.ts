import { Theme } from '@emotion/react';
import { TextProps } from './Text.interfaces';

export function getOrUseCustomColor(
  theme: Theme,
  color: Required<TextProps>['color'],
) {
  if (typeof color !== 'object') return theme.helpers.getColor(color);
  return color.custom;
}
