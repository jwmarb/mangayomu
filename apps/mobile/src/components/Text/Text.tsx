import React from 'react';
import { TextProps as NativeTextProps, Text as NativeText } from 'react-native';
import { variants, colors, alignments } from './styles';
import useStyles from '../../hooks/useStyles';
import { APP_COLORS } from '../types';

export const TEXT_VARIANTS = [
  'h1',
  'h2',
  'h3',
  'body1',
  'body2',
  'button',
] as const;
export const TEXT_COLORS = [
  ...APP_COLORS,
  'textPrimary',
  'textSecondary',
  'disabled',
] as const;
export const TEXT_ALIGNMENTS = ['left', 'center', 'right'] as const;

export type TextVariants = (typeof TEXT_VARIANTS)[number];
export type TextColors = (typeof TEXT_COLORS)[number];
export type TextAlignments = (typeof TEXT_ALIGNMENTS)[number];
export type TextProps = NativeTextProps & {
  variant?: TextVariants;
  color?: TextColors;
  alignment?: TextAlignments;
  contrast?: boolean;
};

/**
 * Text component to render text.
 */
function Text(props: TextProps, ref: React.ForwardedRef<NativeText>) {
  const {
    variant: variantProp = 'body1',
    color: colorProp = 'textPrimary',
    alignment: alignmentProp = 'left',
    contrast = false,
    ...rest
  } = props;
  const color = useStyles(colors, contrast)[colorProp];
  const alignment = alignments[alignmentProp];
  const variant = variants[variantProp];
  const style = [variant, color, alignment, rest.style];
  return <NativeText {...rest} style={style} ref={ref} />;
}

export default React.forwardRef(Text);
