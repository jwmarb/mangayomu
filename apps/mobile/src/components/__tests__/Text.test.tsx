import 'react-native';
import React from 'react';
import Text from '../Text';
import { it } from '@jest/globals';
import renderer from 'react-test-renderer';

jest.mock('react-native-mmkv', () => ({
  useMMKVBoolean: () => [undefined, undefined],
  MMKV: class MMKV {},
}));

jest.mock('../../hooks/useIsDarkMode', () => jest.fn());

it('renders correctly', () => {
  const tree = renderer.create(<Text>Hello World</Text>).toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree).toHaveTextContent('Hello World');
});
