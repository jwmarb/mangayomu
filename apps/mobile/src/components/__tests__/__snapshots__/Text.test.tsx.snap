// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  style={
    [
      {
        "fontSize": 16,
        "fontWeight": "normal",
      },
      {
        "color": "rgba(0, 0, 0, 0.87)",
      },
      {
        "textAlign": "left",
      },
      undefined,
    ]
  }
>
  Hello World
</Text>
`;
